Load event nexus from ILL

MathieuTillet opened the pull request:
**Description of work.**

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->
Make the necessary changes so that ILL's event data can be loaded as event nexuses by Mantid.
Also, in order to start allowing for asked data treatment of it, add an option in GenerateEventsFilter for logarithmic and reverse logarithmic time split, and pass it to the FilterEvents interface. 


**To test:**
- Load `UnitTest/ILL/D22B/042730_trunc.event.nxs`, checking `LoadMonitors`.
- Check there are 1000 events in the workspace.
- Check that the monitor events have been loaded.
- Open the instrument viewer, check the integration boundaries are between 23.71 and 59117.1, and that the instrument looks reasonable.
- Open the FilterEvents interface. Select the workspace, and in the `FilterByTime` tab, check there is a `UseReverseLogarithmic` tick box. Filter by some time interval, check the result is correct. Filter by logarithmic and reverse logarithmic, check the results look sensible (note that reverse logarithmic may not work exactly as expected, due to the way time is handled in `GenerateEventsFilter`. The point is that the bins should expand exponentially starting from the end, not the exact values of this expansion). Check the duration of the created workspaces have been updated to reflect the split.

- Reload the same file after checking `LoadAllLogs`. Check that the sample logs are now filled with a lot of logs. Check that the `start_time` and `end_time` have been correctly loaded. 

- Check quickly that those same steps also work for `ILL/D22B/000242_trunc.event.nxs`.
<!-- Instructions for testing. -->

WIP for #31747 .
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

MathieuTillet committed:
Maek necessary changes for ILL events filtering

ILL event nexuses have a format slightly different from spalliation
sources' ones, so this adapt the code where necessary in order to be
able to read and filter them.

MathieuTillet committed:
Allow loading D22B by name

MathieuTillet committed:
Solve some more warning

MathieuTillet committed:
Fix filtered workspaces warning

When filtering events, the number of workspaces producted is nb of
splitters + 1, because of the workspace for unfiltered events that is
also created.

MathieuTillet committed:
Set correct time limit for filter events

In the filter event interface, use the time from the original workspace
rather than the rebbined one to set the time limits, else some events
are left out

MathieuTillet committed:
Remove proton charge integration when impossible

Event data processing algorithms assume in multiple places that event
data workspaces should have an associated proton charge, when it is not
the case for ILL data.

In those places, whether there is actually a proton charge to integrate
is check before trying to do so, so as to silence multiple unwarranted
warnings.

MathieuTillet committed:
Check if there is something to filter

Before actually trying to. Else throw an error.

MathieuTillet committed:
Update time slider on losing focus

In the filter event interface, when the time start and end fields are
filled using the keyboard, update the slider when the focus is lost or
enter is hit.

MathieuTillet committed:
Harder conditions on slider values

Prevent the max value from being less than the min, which for obvious
reasons should not happen but previously could if the min value was
written in the field

MathieuTillet committed:
Allow for logarithmic binning in filter generation

Add an option in GenerateEventsFilter to allow for logarithmic time
binning, with the same convention as the rebin algorithm.

MathieuTillet committed:
Add logarithmic filter test

Also, a second is 1 000 000 000 of nanoseconds, not one billion and a half
casted to int

MathieuTillet committed:
Add a reverseLogartihm for event filter generation

Add an option to split time intervals by logarithmic length starting
from the end of the filter interval

MathieuTillet committed:
Add test for reverse log filtering

They asked for a bit of contorsion because the offset in the end time of
the run that GenerateEventsFilter adds is very important on the small
value used in the test. But it was determined that the users would
probably never see it, so it wasn't worth dealing with.

MathieuTillet committed:
Add documentation for GenerateEventsFilter

Add documentation for the new options and reformatted a bit the previous
one because the tips are cut short once they run into a dot.

MathieuTillet committed:
Add useReverseLogarithmic to FilterEvents ui

Expose the option in the FilterEvents interface

MathieuTillet committed:
Load all logs in event files

Add an option to load all the logs in a nexus event file, instead of
only those approved by LoadNexusLogs, which are not nearly enough for
ILL's usage.

MathieuTillet committed:
Tests for ill events loader

Add tests for loading events nexus from ILL.

MathieuTillet committed:
Overwrite logs when loading if needed

When loading an event nexus with the LoadAllLogs option checked, some
logs may be loaded twice, thus sending a warning.
It is better to overwrite by default in the general algorithm, as in
most cases there shouldn't be anything to ovrewrite, and in this case,
the version coming from the general algorithm is likely to be better
than the one copied by LoadAllLogs.

MathieuTillet committed:
Add release notes

MathieuTillet committed:
Use smaller test data

MathieuTillet committed:
Remove unused header

gvardany commented on Framework/DataHandling/test/LoadEventNexusTest.h:
```diff
@@ -994,6 +994,61 @@ class LoadEventNexusTest : public CxxTest::TestSuite {
+    TS_ASSERT_DELTA(eventWS->getTofMax(), 13515.0517592763, 1e-2);
+
+    // check that the logs have been loaded by looking at some random example
+    TS_ASSERT_DELTA(eventWS->run().getPropertyAsSingleValue("reactor_power"), 43.21, 1e-2);
```
There is no `turnDown` method in the test suite. You should cleanup the `dummy` workspaces after assertions.

gvardany commented on instrument/Facilities.xml:
```diff
@@ -701,7 +701,9 @@
     <technique>SANS</technique>
   </instrument>
-
+  <instrument name="D22B">
```
Is there a way around without this? D22B is just the upgrade of D22, ideally it should not appear in the facilities file.

MathieuTillet replied:
We have 2 different geometries for the same name, and LoadEventNexus has to decide which one to load based on something.

We could perhaps set the IDF path in the event file directly, and modify LoadEventNexus so that it uses this path if it is provided, but that would mean that for all of ILL event nexuses, the IDF path could not be changed *ever* afterward, which to me looks like a very bad idea.

The only other alternative I see is putting an if clause exclusively for D22 in LoadEventNexus, and it also seems pretty bad.

gvardany replied:
I think the best is to write D22B as instrument name when converting to event.nxs. 
Then we can remove it from the facilities.xml.
Finally, in the load event nexus it can check if the retrieved instrument name exists in the facilities it should do what it does now, otherwise, it can construct the expected IDF path, and run `LoadInstrument` by giving the `Filename` instead of `InstrumentName`. You can check how `LoadILLSANS::getInstrumentFilePath` does it, for example.

gvardany commented:
I still get a warning when loading the files: ```Cannot process ICPevent log. Period 1 assumed for all data.```
In filter events, I get hard errors on putting input like `-0.2` or `20,50000` in the time interval line. Can those be better handled? Is this supposed to be like rebin params or just the step?

MathieuTillet committed:
Ensure workspaces deletion on exiting test

MathieuTillet committed:
Manage errors better in filter events

By not allowing the user to input bad things, and by catching errors
more cleanly.

MathieuTillet committed:
Remove D22B as an instrument

Since it is only a new version, it should not be in the Facilities list,
so it is removed from there, and a new method for finding the instrument
IDF if the instrument is not on this list is implemented in the
LoadEventNexus.

MathieuTillet committed:
Merge branch 'master' into 31747_load_ill_event_data

gvardany commented:
Thanks for the changes. Could you downgrade this warning to information? 
 `Cannot process ICPevent log. Period 1 assumed for all data.`
 
When opening the instrument viewer on the loaded event workspace I get another warning: `overflow encountered in true_divide`. I have a color scale set to log. Do you see that too? 

Also, I'm lost with the time units and what is supposed to go to time interval. Can you post a couple of reasonable inputs (linear, log, reverse log) with expected number of filtered workspaces so that I can test?

gvardany commented on Framework/DataHandling/test/LoadEventNexusTest.h:
```diff
@@ -994,6 +994,65 @@ class LoadEventNexusTest : public CxxTest::TestSuite {
+    TS_ASSERT_THROWS_NOTHING(eventWS = AnalysisDataService::Instance().retrieveWS<EventWorkspace>("dummy"));
+    // Valid WS and it is an EventWorkspace
+    TS_ASSERT(eventWS);
+    if (!eventWS)
```
This return is redundant, if the assertion above fails, the test will end anyhow.

gvardany commented on Framework/DataHandling/test/LoadEventNexusTest.h:
```diff
@@ -994,6 +994,65 @@ class LoadEventNexusTest : public CxxTest::TestSuite {
+    TS_ASSERT_THROWS_NOTHING(eventWS = AnalysisDataService::Instance().retrieveWS<EventWorkspace>("dummy"));
+    // Valid WS and it is an EventWorkspace
+    TS_ASSERT(eventWS);
+    if (!eventWS)
```
same here

MathieuTillet committed:
Remove useless returns in test

MathieuTillet committed:
Lower message importance for ICP

A number of event files don't have ICP time series properties, by
construction. This message thus shouldn't show as a warning, risking to
confuse the user, but stays at information level in case it may be used.

MathieuTillet commented:
I also have the `overflow encountered in true_divide` warning, though it is one I saw more than once and I still don't know where it comes from.

For FilterEvents, I was mistaken : the unit is, as of now, seconds for time interval. For a couple of example, try
 - Linear : `startTime = 0, endTime = 0.059..., TimeInterval = 0.01`, will produce 6 workspaces + 1 (unfiltered)
 - Logarithmic : `startTime = 1e-5, endTime = 0.59..., TimeInterval = -1`, will produce 13 workspaces + 1
 - for ReverseLogarithmic, check the file I sent you, with : `startTime = 1e-5, endTime = 59, TimeInterval = -1`, will produce 6 workspaces + 1

MathieuTillet committed:
Change revLog computation formula

MathieuTillet committed:
Decrease info message from warning

MathieuTillet committed:
Fix issue on time limits in filterevents interface

MathieuTillet commented:
The workspace are reversed ordered (too) in the reverse log case, which is indeed confusing, but might explain what you see.

gvardany commented:
That's what I was wondering. So what's the logic of the workspace order, then? Can we change that? If not, we should at least document it, as it is indeed confusing.

MathieuTillet commented:
The logic is just an effect of how the splits are computed - starting at the end and then going back to the start while increasing the size of the splits. It makes some sense, but we can get rid of this for the sake of consistency with the rest of Mantid.

MathieuTillet committed:
Make sure splitters are correctly ordered

gvardany commented:
With TimeInterval 0.01, and start and stop unchanged (so start empty, end 599), it seems to run into an infinite loop now.

gvardany commented:
With more sensible tests it looks as expected, can you just move the release note to 6.3?

MathieuTillet committed:
Merge branch 'master' into 31747_load_ill_event_data

MathieuTillet committed:
Update release notes

martyngigg commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -247,6 +248,10 @@ void LoadEventNexus::init() {
 
   declareProperty(std::make_unique<PropertyWithValue<bool>>("LoadLogs", true, Direction::Input),
                   "Load the Sample/DAS logs from the file (default True).");
+
```
Could you help me understand the overlap with this new property and `LoadLogs` above - it's not completely clear to me from the diff.

MathieuTillet replied:
If `LoadLogs` is checked, `LoadNexusLogs` will be called as a child algorithm afterward, and then only load a fixed list of logs items (and process some of them) (https://docs.mantidproject.org/nightly/algorithms/LoadNexusLogs-v1.html?highlight=loadnexuslog#data-loaded-from-nexus-file).
On the contrary, `LoadAllLogs` just recursively go through the nexus and load everything that is not data as it is.
Perhaps `LoadRawLogs` would be clearer though ?

martyngigg replied:
Ah I see so it's searching too. Thanks. I'd like to avoid the word Raw as it's an ISIS file format so could get confused with that. Is it worth the parameter having something like Search/Find/Discover in the name? Maybe `SearchForLogs`?

gvardany replied:
I don't think adding search is a good idea, it's more about loading than searching. This is what is done in all ILL loaders: retrieve and load all the metadata. It's striking that LoadLogs option doesn't do that.

martyngigg replied:
I think this is a hangover from sample/run information always being in the same place in SNS/ISIS files.

I'd be happy with `LoadAllLogs` but can we perhaps improve the documentation on both `LoadLogs` and `LoadAllLogs` properties to clarify the difference. As you say people may be surprised that `LoadLogs` doesn't load everything and will wonder where it loads from.

martyngigg commented on Framework/DataHandling/src/LoadEventNexus.cpp:
```diff
@@ -812,27 +819,43 @@ void LoadEventNexus::loadEvents(API::Progress *const prog, const bool monitors)
-        NXevent_data group. Note that the latter is better Nexus compliant.
-      */
-      takeTimesFromEvents = true;
+      prog->doReport("Loading all loads");
```
Should this say `Loading all logs`?

MathieuTillet committed:
Clarify doc on LoadLogs / LoadAllLogs

gvardany reviewed (approved):
thanks for clarifying the documentation

martyngigg commented:
Thanks for updating those bits of documentation.
